crates/core/component/dex/src/component/eviction_manager.rs

```diff
@@ -0,0 +1,112 @@
+    ///    ║  Bottom M  ┃  M+1  │  M+2  │    . . .       │  N-1  │   N   ┃
+    ///    ╚════════════╩━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+    ///     ▽▼▼▼▼▼▼▼▽▽▼▽
+    /// ┌─▶ find overlap
+    /// │   ▲▲▲▲△△△▲▲▲▲▲                                     Q_2: B -> A
+    /// │  ╔════════════╦━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+    /// │  ║  Bottom M  ┃  M+1  │  M+2  │    . . .       │  N-1  │   N   ┃
+    /// │  ╚════════════╩━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+    /// │    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━▶
+    /// │                                                         Ordered by inventory
+    /// │
+    /// │         ▽▼▼▼▼▼▼▼▽▽▼▽
+    /// └───────▶      ∩       Across both indices, the
+    ///           ▲▲▲▲△△△▲▲▲▲▲ bottom M positions overlap
+    ///                        k times where 0 <= k <= M
+    ///                        for N < 2*N_max
+    ///
+    #[instrument(skip_all, err, level = "trace")]
+    async fn evict_positions(&mut self) -> Result<()> {
+        let hot_pairs: BTreeSet<TradingPair> = self.get_active_trading_pairs_in_block();
+        let max_positions_per_pair = self.get_dex_params().await?.max_positions_per_pair;
+
+        for pair in hot_pairs.iter() {
+            let total_positions = self.get_position_count(pair).await;
+            let overhead_size = total_positions.saturating_sub(max_positions_per_pair);
+            if overhead_size == 0 {
+                continue;
+            }
+
+            let pair_ab = DirectedTradingPair::new(pair.asset_1(), pair.asset_2());
+            let pair_ba = pair_ab.flip();
+            let key_ab = eviction_queue::inventory_index::by_trading_pair(&pair_ab);
+            let key_ba = eviction_queue::inventory_index::by_trading_pair(&pair_ba);
+
+            let stream_ab = self.nonverifiable_prefix_raw(&key_ab).boxed();
+            let stream_ba = self.nonverifiable_prefix_raw(&key_ba).boxed();
+
+            let overhead_ab = stream_ab
+                .take(overhead_size as usize)
+                .and_then(|(k, _)| async move {
+                    let raw_id = eviction_queue::inventory_index::parse_id_from_key(k)?;
+                    Ok(position::Id(raw_id))
+                })
+                .try_collect::<BTreeSet<position::Id>>()
+                .await?;
+
+            let overhead_ba = stream_ba
+                .take(overhead_size as usize)
+                .and_then(|(k, _)| async move {
+                    let raw_id = eviction_queue::inventory_index::parse_id_from_key(k)?;
+                    Ok(position::Id(raw_id))
+                })
+                .try_collect::<BTreeSet<position::Id>>()
+                .await?;
+
+            let overlap = overhead_ab.intersection(&overhead_ba);
+
+            for id in overlap {
+                self.close_position_by_id(id).await?;
+            }
+        }
+        Ok(())
+    }
```

Comment:
I considered `JoinSet`ifying across trading pairs, but it feels like unlikely to be helpful because it would take a great deal of trading pairs for this code to be slow, and gas costs and ineluctable eviction would make it exorbitant to sustain.

Reply:
that makes sense to me! :+1: